docs/content/2.deploy/providers/firebase.md

```diff
@@ -82,11 +104,86 @@ NITRO_PRESET=firebase yarn build
 firebase emulators:start
 ```
 
-## Deploy to Firebase Hosting via CLI
+## Build and Deploy
 
-Deploying to Firebase Hosting is a simple matter of just running the `firebase deploy` command.
+Deploy to Firebase Hosting by running a nitro build and then running the `firebase deploy` command.
 
 ```bash
 NITRO_PRESET=firebase yarn build
+```
+
+```bash
+npx firebase-tools deploy
+```
+
+If you chose the alternative setup using the Firebase CLI, you can also run:
+
+```bash
 firebase deploy
 ```
+
+## Using 2nd Generation Firebase Functions
+
+- [Comparison between 1st and 2nd generation functions](https://firebase.google.com/docs/functions/version-comparison)
+
+To switch to the more recent and, recommended generation of firebase functions, set the `FIREBASE_FUNCTIONS_GEN` to `2` when building:
+
+```bash
+FIREBASE_FUNCTIONS_GEN=2 NITRO_PRESET=firebase yarn build
+```
+
+If you already have a deployed version of your website and want to upgrade to 2nd gen, [see the Migration process on Firebase docs](https://firebase.google.com/docs/functions/2nd-gen-upgrade). Namely, the CLI will ask you to delete your existing functions before deploying the new ones.
+
+## Options
+
+You can set options for the firebase functions in your `nitro.config.ts` file:
+
+```ts
+import { defineNitroConfig } from 'nitropack/config'
+
+export default defineNitroConfig({
+  // Nitro options...
+  firebase: {
+    gen2: {
+      httpOptions: {
+        region: "europe-west1",
+        maxInstances: 3,
+        // etc.
+      }
+    }
+  }
+})
+```
+
+You can also set 1st gen options in the `gen1` key.
+
+## Runtime
+
+Firebase uses the `engines.node` version in your `package.json` to determine which node version to use for your functions. If you want to use a different version of node for your functions, you should set the `engines.node` version in your `package.json` to the version you want to use:
+
+```json [package.json]
+{
+  "engines": {
+    "node": "18"
```

Comment:
This field is also used for Node version of building nitro app itself. Can't we purely depend on `firebase.json` to configure runtime node? What is purpose of defined node version here? (if we really need, it should be _generated_ in `.output/package.json` and picked somehow by cli)

Reply:
It's not mandatory, I have projects where I don't set it. This part was already added by @JamieCurnow so they will probably know better

Reply:
Okay let's remove it then to simplify and then iterate over to somehow introduce if needed

Reply:
@pi0 this is used by firebase to determine the runtime of node to run the cloud function in. It is required as far as I know / as far as I can tell from the docs

https://firebase.google.com/docs/functions/manage-functions?gen=2nd#set_nodejs_version

> The engines field is required; it must specify one of the supported Node.js versions in order for you to deploy and run functions

Reply:
The nitro preset for firebase will default to node 18 and warn of incompatible versions

https://github.com/unjs/nitro/pull/1500/files/93d49aab552a9d22449d581045047da97213edb8#diff-10ef001537e49ef8e8387ba0cf14acbdbcef1cc58a1aebfad7e93e28073c1274R81

Reply:
From docs:

> Set the version in the engines field in the package.json file that was created in your functions/ directory during initialization. 

I think we can generate `package.json` with desired engines field in `.output/` directory to specify (runtime) node.

Reply:
@pi0 Yea it definitely reads the generated package.json in the output directory. I guess we could introduce it as a preset option instead of reading it from the root package.json as it does now?

Reply:
Preset option for runtime node seems nice idea 💯  (still, i am wondering is it a required option as posva mentioned seems not and do we want to overide firebase defaults? it has some benefits like knowing our build target..)

Reply:
@pi0 it would make sense to me that it should be the same node version that nitro uses to build. Do we have access to that version at all?

Reply:
I'm guessing a developer may experience issues if nitro builds on node 18 and firebase deploys on node 14 for example 😅

Reply:
It doesn't matter what (latest) version of node we use. But it is important what transforms `esbuild` applies to the `.output` files as that matters and that would be the main benefit of knowing runtime node version (in `firebase.nodeVersion`) for possible syntax transforms and polyfills when needed.

Reply:
I think it's better to keep this instruction. One can also set the `functions.runtime` option in `firebase.json` but that seems less documented and it's just more convenient setting it in package.json.

Reply:
Build time Node.js version !== Runtime Node.js version, so no i don't think it is a good idea to document it this way.

Reply:
In the case of firebase projects, it is, since you run emulators using a specific node of version locally.